File: src/main/scala/nl.knaw.dans.easy.dd2d/DatasetEditor.scala
/*
 * Copyright (C) 2022 DANS - Data Archiving and Networked Services (info@dans.knaw.nl)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package nl.knaw.dans.easy.dd2d

import nl.knaw.dans.easy.dd2d.mapping.{ AccessRights, License }
import nl.knaw.dans.easy.dd2d.migrationinfo.BasicFileMeta
import nl.knaw.dans.lib.dataverse.DataverseClient
import nl.knaw.dans.lib.error.{ TraversableTryExtensions, TryExtensions }
import nl.knaw.dans.lib.logging.DebugEnhancedLogging
import nl.knaw.dans.lib.scaladv.model.dataset.Embargo
import nl.knaw.dans.lib.scaladv.model.file.FileMeta
import nl.knaw.dans.lib.scaladv.model.file.prestaged.PrestagedFile
import org.json4s.native.Serialization
import org.json4s.{ DefaultFormats, Formats }

import java.net.URI
import java.nio.file.{ Path, Paths }
import java.util.regex.Pattern
import java.util.{ Date, Optional }
import scala.util.{ Failure, Success, Try }

/**
 * Object that edits a dataset, a new draft.
 */
abstract class DatasetEditor(dataverseClient: DataverseClient, optFileExclusionPattern: Option[Pattern], zipFileHandler: ZipFileHandler) extends DebugEnhancedLogging {
  type PersistentId = String
  type DatasetId = Int
  implicit val jsonFormats: Formats = DefaultFormats

  /**
   * Performs the task.
   *
   * @return the persistentId of the dataset created or modified
   */
  def performEdit(): Try[PersistentId]

  protected def addFiles(persistentId: String, files: List[FileInfo], prestagedFiles: Set[BasicFileMeta] = Set.empty): Try[Map[Int, FileInfo]] = Try {
    trace(persistentId, files)
    files.map { f =>
      debug(s"Adding file, directoryLabel = ${ f.metadata.directoryLabel }, label = ${ f.metadata.label }")
      val id = addFile(persistentId, f, prestagedFiles).unsafeGetOrThrow
      dataverseClient.dataset(persistentId).awaitUnlock()
      id -> f
    }.toMap
  }

  private val noFile: java.io.File = null

  private def addFile(doi: String, fileInfo: FileInfo, prestagedFiles: Set[BasicFileMeta]): Try[Int] = {
    val result = for {
      r <- getPrestagedFileFor(fileInfo, prestagedFiles).map { prestagedFile =>
        logger.info(s"Adding prestaged file (scala -> java change not tested): $fileInfo") // TODO
        Try(dataverseClient.dataset(doi).addFileItem(
          Optional.of(noFile),
          Optional.of(Serialization.write(prestagedFile))
        ))
      }.getOrElse {
        debug(s"Uploading file: $fileInfo")
        val optWrappedZip = zipFileHandler.wrapIfZipFile(fileInfo.file)
        val r = Try(dataverseClient.dataset(doi).addFileItem(
          Optional.of(optWrappedZip.getOrElse(fileInfo.file).toJava), // TODO what about closing files?
          Optional.of(Serialization.write(fileInfo.metadata))
        ))
        optWrappedZip.foreach(_.delete(swallowIOExceptions = true))
        r
      }
      files <- Try(r.getData)
      triedId = Try(files.getFiles.get(0).getDataFile.getId)
      _ <- Try(dataverseClient.dataset(doi).awaitUnlock())
    } yield triedId
    debug(s"Result = $result")
    result.map(_.getOrElse(throw new IllegalStateException("Could not get DataFile ID from response")))
  }

  protected def getPathToFileInfo(deposit: Deposit): Try[Map[Path, FileInfo]] = {
    for {
      bagPathToFileInfo <- deposit.getPathToFileInfo
      pathToFileInfo = bagPathToFileInfo.map { case (bagPath, fileInfo) => Paths.get("data").relativize(bagPath) -> fileInfo }
      filteredPathToFileInfo = excludeFiles(pathToFileInfo)
    } yield filteredPathToFileInfo
  }

  private def excludeFiles(p2fi: Map[Path, FileInfo]): Map[Path, FileInfo] = {
    trace(p2fi)
    p2fi.toList.filter {
      case (p, _) =>
        val foundMatch = optFileExclusionPattern.forall(_.matcher(p.toString).matches())
        if (foundMatch) logger.info(s"Excluding file: ${ p.toString }")
        !foundMatch
    }.toMap
  }

  protected def getPrestagedFileFor(fileInfo: FileInfo, basicFileMetas: Set[BasicFileMeta]): Option[PrestagedFile] = {
    val matchingChecksums = basicFileMetas.filter(_.prestagedFile.checksum.`@value` == fileInfo.checksum)
    if (matchingChecksums.size == 1) Option(matchingChecksums.head.prestagedFile)
    else if (matchingChecksums.isEmpty) Option.empty // no matches
         else { // multiple matches
           val matchingPaths = basicFileMetas.filter(bfm => bfm.label == fileInfo.metadata.label.get && bfm.directoryLabel == fileInfo.metadata.directoryLabel)
           if (matchingPaths.size == 1) Option(matchingPaths.head.prestagedFile)
           else if (matchingPaths.isEmpty) Option.empty
                else throw new IllegalArgumentException("Found multiple basic file metas with the same path in a single dataset version")
         }
  }

  protected def updateFileMetadata(databaseIdToFileInfo: Map[Int, FileMeta]): Try[Unit] = {
    trace(databaseIdToFileInfo)
    databaseIdToFileInfo.map { case (id, fileMeta) =>
      val json = Serialization.write(fileMeta)
      debug(s"id = $id, json = $json")
      val r = Try(dataverseClient.file(id).updateMetadata(json))
      debug(s"id = $id, result = $r")
      r
    }.collectResults.map(_ => ())
  }

  protected def configureEnableAccessRequests(deposit: Deposit, persistendId: PersistentId, canEnable: Boolean): Try[Unit] = {
    for {
      ddm <- deposit.tryDdm
      files <- deposit.tryFilesXml
      enable = AccessRights.isEnableRequests((ddm \ "profile" \ "accessRights").head, files)
      _ = logger.trace("AccessRequests enable "+ enable + " can " +canEnable)
      _ <- if (!enable) Try(dataverseClient.accessRequests(persistendId).disable())
           else if (canEnable) Try(dataverseClient.accessRequests(persistendId).enable())
                else Success(())
    } yield ()
  }

  protected def licenseAsJson(supportedLicenses: List[URI])(variantToNormalized: Map[String, String])(deposit: Deposit): Try[String] = {
    trace(deposit)
    for {
      ddm <- deposit.tryDdm
      license <- (ddm \ "dcmiMetadata" \ "license").find(License.isLicenseUri).map(Success(_)).getOrElse(Failure(RejectedDepositException(deposit, "No license specified")))
      uri = License.getLicenseUri(supportedLicenses)(variantToNormalized)(license).toASCIIString
    } yield s"""
                |{ "http://schema.org/license": "$uri" }
                |""".stripMargin
  }

  protected def isEmbargo(date: Date): Boolean = {
    date.compareTo(new Date()) > 0
  }

  protected def embargoFiles(persistendId: PersistentId, dateAvailable: Date, fileIds: List[Int]): Try[Unit] = {
    trace(persistendId, fileIds)
    val embargo = Embargo(dateAvailableFormat.format(dateAvailable), "", fileIds)
    val json = Serialization.write(embargo)
    Try(dataverseClient.dataset(persistendId).setEmbargo(json))
  }

  protected def deleteDraftIfExists(persistentId: String): Unit = {
    val result = for {
      v <- Try(dataverseClient.dataset(persistentId).viewLatestVersion().getData)
      _ = logger.trace("deleting draft")
      _ <- if (v.getLatestVersion.getVersionState.contains("DRAFT"))
             deleteDraft(persistentId)
           else Success(())
    } yield ()
    result.doIfFailure {
      case e => logger.warn("Could not delete draft", e)
    }
  }

  private def deleteDraft(persistentId: PersistentId): Try[Unit] = {
    for {
      _ <- Try(dataverseClient.dataset(persistentId).deleteDraft())
      _ = logger.info(s"DRAFT deleted")
    } yield ()
  }
}

File: src/main/java/nl/knaw/dans/ingest/core/service/DatasetEditor.java
/*
 * Copyright (C) 2022 DANS - Data Archiving and Networked Services (info@dans.knaw.nl)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package nl.knaw.dans.ingest.core.service;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import nl.knaw.dans.easy.dd2d.Deposit;
import nl.knaw.dans.easy.dd2d.RejectedDepositException;
import nl.knaw.dans.lib.dataverse.DataverseClient;
import nl.knaw.dans.lib.dataverse.DataverseException;
import nl.knaw.dans.lib.dataverse.Version;
import nl.knaw.dans.lib.dataverse.model.dataset.Dataset;
import nl.knaw.dans.lib.dataverse.model.dataset.Embargo;
import org.apache.commons.lang.ArrayUtils;
import scala.collection.JavaConverters;
import scala.xml.Node;

import java.io.IOException;
import java.net.URI;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Collection;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

@Slf4j
public abstract class DatasetEditor {

    protected final DataverseClient dataverseClient;
    protected final boolean isMigration;
    protected final Dataset dataset;
    protected final Deposit deposit;
    protected final Map<String, String> variantToLicense;
    protected final List<URI> supportedLicenses;

    protected final int publishAwaitUnlockMillisecondsBetweenRetries;
    protected final int publishAwaitUnlockMaxNumberOfRetries;

    protected final Pattern fileExclusionPattern;
    protected final ZipFileHandler zipFileHandler;

    protected final ObjectMapper objectMapper;
    private final DateFormat dateAvailableFormat = new SimpleDateFormat("yyyy-MM-dd");

    protected DatasetEditor(DataverseClient dataverseClient,
        boolean isMigration,
        Dataset dataset,
        Deposit deposit,
        Map<String, String> variantToLicense,
        List<URI> supportedLicenses,
        int publishAwaitUnlockMillisecondsBetweenRetries,
        int publishAwaitUnlockMaxNumberOfRetries, Pattern fileExclusionPattern, ZipFileHandler zipFileHandler, ObjectMapper objectMapper) {
        this.dataverseClient = dataverseClient;
        this.isMigration = isMigration;
        this.dataset = dataset;
        this.deposit = deposit;
        this.variantToLicense = variantToLicense;
        this.supportedLicenses = supportedLicenses;
        this.publishAwaitUnlockMillisecondsBetweenRetries = publishAwaitUnlockMillisecondsBetweenRetries;
        this.publishAwaitUnlockMaxNumberOfRetries = publishAwaitUnlockMaxNumberOfRetries;
        this.fileExclusionPattern = fileExclusionPattern;
        this.zipFileHandler = zipFileHandler;
        this.objectMapper = objectMapper;
    }

    public abstract String performEdit() throws IOException, DataverseException, InterruptedException;

    Map<Integer, FileInfo> addFiles(String persistentId, Collection<FileInfo> fileInfos) throws IOException, DataverseException {
        var result = new HashMap<Integer, FileInfo>(fileInfos.size());

        for (var fileInfo : fileInfos) {
            log.debug("Adding file, directoryLabel = {}, label = {}",
                fileInfo.getMetadata().getDirectoryLabel(), fileInfo.getMetadata().getLabel());

            var id = addFile(persistentId, fileInfo);
            result.put(id, fileInfo);
        }

        return result;
    }

    private Integer addFile(String persistentId, FileInfo fileInfo) throws IOException, DataverseException {
        var dataset = dataverseClient.dataset(persistentId);
        var wrappedZip = zipFileHandler.wrapIfZipFile(fileInfo.getPath());

        var file = wrappedZip.orElse(fileInfo.getPath());
        var metadata = objectMapper.writeValueAsString(fileInfo.getMetadata());
        var result = dataset.addFileItem(Optional.of(file.toFile()), Optional.of(metadata));

        if (wrappedZip.isPresent()) {
            try {
                Files.deleteIfExists(wrappedZip.get());
            }
            catch (IOException e) {
                log.error("Unable to delete zipfile {}", wrappedZip.get(), e);
            }
        }

        log.debug("Result: {}", result);
        return result.getData().getFiles().get(0).getDataFile().getId();
    }

    protected String getLicense(Node node) {
        // TODO get license from XPath expression
        var licenseNode = node.$bslash("dcmiMetadata").$bslash("license")
            .find(License::isLicenseUri);

        if (licenseNode.isEmpty()) {
            throw new RejectedDepositException(deposit, "no license specified", null);
        }

        return License.getLicenseUri(supportedLicenses, variantToLicense, licenseNode).toASCIIString();
    }

    protected String toJson(Map<String, String> input) throws JsonProcessingException {
        return objectMapper.writeValueAsString(input);
    }

    Map<Path, FileInfo> getFileInfo() {
        return JavaConverters.mapAsJavaMap(deposit.getPathToFileInfo().get())
            .entrySet().stream()
            .map(entry -> {
                // conver to new FileInfo class
                var f = entry.getValue();
                var fileInfo = new FileInfo(f.file().toJava().toPath(), f.checksum(), f.metadata());

                return Map.entry(entry.getKey(), fileInfo);
            })
            .map(entry -> {
                // relativize the path
                var bagPath = entry.getKey();
                var fileInfo = entry.getValue();
                var newKey = Paths.get("data").relativize(bagPath);

                return Map.entry(newKey, fileInfo);
            })
            .filter(entry -> {
                // remove entries that match the file exclusion pattern
                var path = entry.getKey().toString();
                if (fileExclusionPattern != null) {
                    return !fileExclusionPattern.matcher(path).matches();
                }
                return true;
            })
            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
    }

    void embargoFiles(String persistentId, Date dateAvailable) throws IOException, DataverseException {
        if (!dateAvailable.after(new Date())) {
            log.debug("Date available in the past, no embargo: {}", dateAvailable);
        }
        else {
            var api = dataverseClient.dataset(persistentId);
            var files = api.getFiles(Version.LATEST.toString()).getData();

            var items = files.stream()
                .filter(f -> !"easy-migration".equals(f.getDirectoryLabel()))
                .map(f -> f.getDataFile().getId())
                .collect(Collectors.toList());

            embargoFiles(persistentId, dateAvailable, items);
        }
    }

    void embargoFiles(String persistentId, Date dateAvailable, Collection<Integer> fileIds) throws IOException, DataverseException {
        if (!dateAvailable.after(new Date())) {
            log.debug("Date available in the past, no embargo: {}", dateAvailable);
        }
        else {
            var api = dataverseClient.dataset(persistentId);
            var embargo = new Embargo(dateAvailableFormat.format(dateAvailable), "",
                ArrayUtils.toPrimitive(fileIds.toArray(Integer[]::new)));

            api.setEmbargo(embargo);
            api.awaitUnlock(publishAwaitUnlockMaxNumberOfRetries, publishAwaitUnlockMillisecondsBetweenRetries);
        }
    }
}
